safe_transaction_service/account_abstraction/models.py

```diff
@@ -83,6 +84,24 @@ def to_user_operation(self, add_tx_metadata: bool = False) -> UserOperationClass
             else None
         )
 
+        if self.signature:
+            signature = HexBytes(self.signature)
+        elif self.safe_operation:
+            # `UserOperation` `signature` is used to store data for the `SafeOperation`
+            # First 6 bytes of the signature are `SafeOperation` `valid_after` and next 6 are `valid_until`
+            # Real user signature is after those 12 bytes
+            # If there's no signature it should be built from `SafeOperation` fields
+            valid_after, valid_until = [
+                int(valid_date.timestamp()) if valid_date else 0
+                for valid_date in (
+                    self.safe_operation.valid_after,
+                    self.safe_operation.valid_until,
+                )
+            ]
+            signature = encode_packed(["uint48"] * 2, [valid_after, valid_until])
+        else:
+            signature = b""
```

Comment:
As discussed, signatures should not be empty and probably we can throw here instead 👍